lib/travis/api/app/services/schedule_request.rb

```diff
@@ -40,13 +58,45 @@ def payload
         end
 
         def repo
-          @repo ||= Repository.by_slug(slug).first
+          instance_variable_defined?(:@repo) ? @repo : @repo = Repository.by_slug(slug).first
         end
 
         def slug
           repo = params[:repository] || {}
           repo.values_at(:owner_name, :name).join('/')
         end
+
+        def throttle
+          @throttle ||= Throttle.new(slug)
+        end
+
+        class Throttle < Struct.new(:slug)
+          def throttled?
+            current_requests >= max_requests
+          end
+
+          def message
+            'API throttled'
+          end
+
+          private
+
+            def current_requests
+              @current_requests ||= begin
+                sql = 'repository_id = ? AND result = ? AND created_at > ?'
```

Comment:
@joshk you're right that indexes need to be checked. i'll "explain" the query later. i was hoping that due to the sorting and limit the query should be pretty fast, but i don't actually know.